### PR TITLE
Convert to TypeScript

### PR DESCRIPTION
**Pull request by charpeni**
As this library is a client interface to Clubhouse's API, we could say everything is about typing and requests.

In order to support proper typings, let's try out TypeScript and from there we ~— should — be able to~ generate Flow types.

Things to do:

- [x] Setup TypeScript
- [x] Convert the lib to TS
- [x] Convert tests to TS
- [x] Configure CircleCI
- [x] Test it
- [x] Generate Flow types using [Flowgen](https://github.com/joarwilk/flowgen)

Flow files were converted with a codemod tool [flow2ts](https://github.com/bcherny/flow-to-typescript), when it couldn't be done automatically, I did it manually with the help of [flow-to-ts](https://flow-to-ts.netlify.com/).

### CODE REVIEW

**Comment by charpeni** [comment]
@j-martin let me know what you think!

Looking to merge those conflicts.

File: src/index.ts [hunk 1]
```diff
@@ -67,8 +65,8 @@ class Client<RequestType, ResponseType> {
   /** */
   static create(
     token: string,
-    config?: ClientConfig = defaultConfig,
-  ): Client<Request, Response> {
+    config: ClientConfig = defaultConfig,
+  ): Client<any, any> {
```
Comment (charpeni) [inline src/index.ts]: We should be able to use `Request` and `Response` from the DOM lib.

File: scripts/generate-flow-types.sh [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# /bin/sh
+
+for file in $(find ./lib -name *.d.ts -type f); do flowgen --add-flow-header ${file} -o ${file/.d.ts/.js.flow}; done;
```
Comment (j-martin) [inline scripts/generate-flow-types.sh]: I would rewrite this like that:
```
#!/usr/bin/env bash

set -o errexit
set -o pipefail
set -u

readonly IFS=$'\n\t'

_main() {
  for file in $(find ./lib -name *.d.ts -type f); do
    flowgen --add-flow-header "${file}" -o "${file/.d.ts/.js.flow}"
  done;
}

_main "$@"

```

In its current (the PR) form:
- flowgen will continue to process files even if there is an error (and the error will be masked). 
- File names with space could be an issue (won't be an issue, but shellcheck does not like it).
- That's a long line.

File: package.json [hunk 3]
```diff
@@ -12,7 +12,7 @@
   "scripts": {
     "build": "rimraf ./lib && yarn build:typescript && yarn build:flow",
     "build:typescript": "tsc",
-    "build:flow": "for file in $(find ./lib -name *.d.ts -type f); do flowgen --add-flow-header ${file} -o ${file/.d.ts/.js.flow}; done; node scripts/convert-import-from-types",
+    "build:flow": "bash scripts/generate-flow-types.sh && node scripts/convert-import-from-types",
```
Comment (j-martin) [inline package.json]: `chmod +x scripts/generate-flow-types.sh`

and `"build:flow": "scripts/generate-flow-types.sh && node scripts/convert-import-from-types",` instead.

File: scripts/generate-flow-types.sh (hunk 2 above)
Comment (charpeni) [reply scripts/generate-flow-types.sh]: Sounds good! Thanks for reviewing this bash file, I was holding myself not to rewrite it in JS. 😅 

Thanks to you for leading the way to TypeScript.

**Comment by charpeni** [comment]
@j-martin The new version has been released as `0.9.0-rc.0`, let us know what you think!

**Pull request merged**